### ISSUE TITLE
Crest based processes will sometimes not recovery properly if Cassandra fails

### ISSUE THREAD

**Issue by mirw** [Metaswitch/crest#283]
[AME2: I have changed the title of this issue. It did previously indicate a issue with Homer, however Homer seems to be behaving as expected (see the logs I posted below).]
#### Symptoms

Start an all-in-one image, log in through Ellis and create an account.  The web account creates successfully, but the SIP account fails.  It appears that Homer has locked up.

After I restart Homer, everything works fine.  Interestingly, it also creates a log file dated an hour earlier than the previous one (and correct according to UTC) - maybe this is somehow related to DST?
#### Impact

Homer locks up until it's manually restarted.  Can't create subscribers through Ellis or provide call service through Sprout.
#### Release and environment

Release-98, All-in-One Image on VirtualBox.
#### Steps to reproduce

Start up and All-in-One image.  Seems 100% reproducible at present.

**Comment by mirw** [comment]
OK, I hit this 2/2 times last night, and now it won't repro this morning (4/4 times).  I still think we should investigate fairly urgently, though.

**Comment by eleanor-merry** [comment]
It's strange that monit doesn't restart us in this situation. Should we be looking to beef up our monit polling? E.g. GET a simservs XML and check we get a 200/404 rather than just pinging Homer

**Comment by mirw** [comment]
Just repro-ed this again (I think).

I wonder whether I was wrong to say this was homer locking up - it may just be that it doesn't log what it's doing properly.  Most recently, it looked like Cassandra was running very slowly - it took several minutes of running at 100% CPU before it became operational.

**Comment by mirw** [comment]
OK, so it looks like there are a couple of issues here.

*   Homer (and maybe Homestead-prov?) don't seem to give any error indication if they fail to communicate with Cassandra (maybe specifically if they time out - I need to do some more digging on this).

*   All-in-one images take **much** longer than they used to to get up and running.  I think we used to have a pre-built Cassandra cluster with pre-installed schema in our AIO images, but it looks like we use etcd to cluster us (which takes at least 30s) and then have to wait for Cassandra to start running before we can inject the schema and provide service.

**Comment by mirw** [comment]
`/var/log/homer/homer-err.log` reveals that:

```
Unhandled error in Deferred:


Traceback (most recent call last):
Failure: twisted.internet.error.ConnectionRefusedError: Connection was refused by other side: 111: Connection refused.
```

Looking at `PassthroughHandler`, it looks like it doesn't handle any exceptions - I think we should add that.

**Comment by mirw** [comment]
Hmm... actually, it looks like `PassthroughHandler` isn't the class that receives the exception - it seems that CassandraClient queues requests while it is disconnected, and it looks like ManagedCassandraClientFactory will try to reconnect regularly.

Here was my test app:

```
from telephus.client import CassandraClient
from telephus.protocol import ManagedCassandraClientFactory
from twisted.internet import reactor

def log(msg):
    def func(x):
        print msg
    return func

factory = ManagedCassandraClientFactory("homer")
reactor.connectTCP("127.0.0.1", 9160, factory)
cass = CassandraClient(factory)
cass.get(column_family="simservs", key="1234566", column="value").addCallback(log("get OK")).addErrback(log("get failed"))
reactor.run()
```

If I run this with Cassandra stopped, I get no callback or errback from the cass.get call.  Fairly soon after I start Cassandra, I **do** get a callback.  I think we need to look at whether CassandraClient has a "connected" property.

**Comment by mirw** [comment]
Apparently, there's no "connected" status on CassandraClient or ManagedCassandraClientFactory.  I think we'd need to add callbacks to track this, and not attempt calls into CassandraClient if the connection to Cassandra is not up.  This is going somewhat beyond the original issue, though, so maybe this aspect is no longer high priority.

**Comment by chris-elford-metaswitch** [comment]
We have a few options:

1. Figure out why we are failing to connect to Cassandra when it comes up, and fix it.

2. Add some sort of logic to either Crest or telephus to time out requests after a while. This is likely to be tricky and prone to errors.

3. Go with the suggestion from @eleanor-merry and enhance our ping script.

Given that Homer isn't on the call path, I think 3 is the best chance of avoiding persistent outages. I'm going to look into doing that.

**Comment by chris-elford-metaswitch** [comment]
Rob's corrected me, Homer is actually an XDMS, so it is on the call path. Still, option 3 looks best to get this working again in the short term.

**Comment by chris-elford-metaswitch** [comment]
I've tested a fix to get monit to restart Homer when it locks up.

**Comment by eleanor-merry** [comment]
Reopening as I've seen the same issue with the new release

**Comment by rkd-msw** [comment]
Looking at https://github.com/Metaswitch/crest/pull/287, that was tested by stopping Cassandra while Homer was running, whereas I think the failure case here is Homer coming up before Cassandra has - maybe that behaves differently?

**Comment by rkd-msw** [comment]
Reopening - I've seen this again. I'll copy diags on shortly.

**Comment by rkd-msw** [comment]
Diags attached:

* a homestead-prov log showing requests being received and never responded to
* a homestead-prov log showing the telltale "twisted.internet.error.ConnectionRefusedError: Connection was refused by other side: 111: Connection refused."
* monit.log showing that we didn't detect ping failures and restart Homestead

[monit.log.txt](https://github.com/Metaswitch/crest/files/511241/monit.log.txt)
[homestead-prov_20160927T140000Z.txt](https://github.com/Metaswitch/crest/files/511242/homestead-prov_20160927T140000Z.txt)
[homestead-prov-err.log.txt](https://github.com/Metaswitch/crest/files/511304/homestead-prov-err.log.txt)

**Comment by AME2** [comment]
I am also seeing the telltale "twisted.internet.error.ConnectionRefusedError: Connection was refused by other side: 111: Connection refused." log in homer-err.log.old when i start up an All in One node.

**Comment by AME2** [comment]
Having looked through the attached monit.log there does not appear to be any issues with Homer. Whenever Homer fails it ping twice without intermediate successes monit will attempt to restart the Homer process:
UTC Oct  4 23:52:08] error    : 'poll_homer' '/usr/share/clearwater/bin/poll_homer.sh' failed with exit status (1) -- HTTP failed to http://127.0.0.1:7888/ping
[UTC Oct  4 23:52:08] info     : 'cassandra_process' process is running with pid 32358
[UTC Oct  4 23:52:18] error    : 'poll_homer' '/usr/share/clearwater/bin/poll_homer.sh' failed with exit status (1) -- HTTP failed to http://127.0.0.1:7888/ping
[UTC Oct  4 23:52:18] info     : 'poll_homer' exec: /etc/init.d/homer
However @rkd-msw has seen under this issue a problem where subscribers cannot be registered unless Homestead-prov is restarting manually. I will rename this issue to reflect this.
I have failed to reproduce this problem after importing five instances of the all-in-one image into VirtualBox, I suggest that when cutting a release that people keep an eye out for this issue.
[monit.log.txt](https://github.com/Metaswitch/crest/files/543983/monit.log.txt)

**Comment by AME2** [comment]
In attempting to reproduce this issue I have raised another issue which has very similar symptoms (being unable to register subscribers on an all-in-one node), see [Ellis iss#192](https://github.com/Metaswitch/ellis/issues/192).
I have also seen as part of debugging this issue that Homestead-prov and Homer share the same PingHandler class but interact with it very differently which can be a little confusing, see iss #310

**Comment by AlexHockey** [comment]
The reason we're still seeing a problem on homestead is that the ping handler finds out what cassandra clients to test by asking the `PassthroughHandler`, which Homestead just doesn't use. I'm going to refactor this code so that applications give the ping handler a way to get the clients that are actually being used.

**Comment by eleanor-merry** [comment]
@AlexHockey - is this fixed now? (excluding https://github.com/Metaswitch/crest/issues/313)

**Comment by AlexHockey** [comment]
Yes - the node will eventually recover. 

I'm not happy that the cassandra client does not automatically try to reconnect and resend requests, but I think we'll probably end up looking at that when we support cassandra running on a separate node, so it's not worth keeping this issue open.
